// @lovable.dev/vite-plugin-hmr-gate/dist/index.js
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
//#region src/index.ts
function hmrGatePlugin(options = {}) {
	const fullReload = options.fullReload ?? true;
	const gatedEvents = new Set(options.events ?? [
		"change",
		"add",
		"unlink"
	]);
	const passthroughPatterns = options.passthrough ?? [];
	const pending = /* @__PURE__ */ new Map();
	const lastFlushedHashes = /* @__PURE__ */ new Map();
	let originalEmit = null;
	let server = null;
	function fileContentHash(filePath) {
		try {
			return createHash("sha1").update(readFileSync(filePath)).digest("hex");
		} catch {
			return null;
		}
	}
	function isPassthrough(filePath) {
		if (typeof filePath !== "string") return true;
		if (filePath.includes("node_modules/.vite")) return true;
		return passthroughPatterns.some((p) => filePath.includes(p));
	}
	return {
		name: "hmr-gate",
		apply: "serve",
		configureServer(srv) {
			server = srv;
			const _emit = srv.watcher.emit.bind(srv.watcher);
			originalEmit = _emit;
			srv.watcher.emit = ((event, ...args) => {
				if (typeof event === "string" && gatedEvents.has(event) && !isPassthrough(args[0])) {
					const filePath = args[0];
					if (event === "change" || event === "add") {
						const hash = fileContentHash(filePath);
						if (hash !== null && hash === lastFlushedHashes.get(filePath)) {
							pending.delete(filePath);
							return false;
						}
					}
					let events = pending.get(filePath);
					if (!events) {
						events = /* @__PURE__ */ new Set();
						pending.set(filePath, events);
					}
					events.add(event);
					return false;
				}
				return _emit(event, ...args);
			});
			srv.middlewares.use((req, res, next) => {
				if (req.method !== "POST" || req.url !== "/__hmr_flush") return next();
				const entries = Array.from(pending.entries());
				pending.clear();
				for (const [f] of entries) srv.moduleGraph.onFileChange(f);
				if (entries.length > 0) {
					for (const [f, events] of entries) {
						for (const event of events) _emit(event, f);
						const hash = fileContentHash(f);
						if (hash === null) lastFlushedHashes.delete(f);
						else lastFlushedHashes.set(f, hash);
					}
					if (fullReload) srv.ws.send({ type: "full-reload" });
				}
				const files = entries.map(([f]) => f);
				const body = JSON.stringify({
					flushed: files,
					count: entries.length,
					mode: fullReload ? "full-reload" : "granular"
				});
				res.writeHead(200, { "Content-Type": "application/json" });
				res.end(body);
			});
			srv.middlewares.use((req, res, next) => {
				if (req.method !== "GET" || req.url !== "/__hmr_gate") return next();
				const serialized = {};
				for (const [f, events] of pending) serialized[f] = Array.from(events);
				const body = JSON.stringify({
					enabled: true,
					pending: serialized,
					count: pending.size,
					mode: fullReload ? "full-reload" : "granular"
				});
				res.writeHead(200, { "Content-Type": "application/json" });
				res.end(body);
			});
		},
		closeWatcher() {
			if (server && originalEmit) server.watcher.emit = originalEmit;
			pending.clear();
			lastFlushedHashes.clear();
			server = null;
			originalEmit = null;
		}
	};
}
//#endregion
export { hmrGatePlugin };
